backend/test_observer/controllers/test_executions/end_test.py

```diff
@@ -0,0 +1,181 @@
+from fastapi import APIRouter, Depends, HTTPException
+from sqlalchemy import select
+from sqlalchemy.orm import Session, joinedload
+
+from test_observer.data_access.models import (
+    Artefact,
+    ArtefactBuild,
+    Stage,
+    TestCase,
+    TestExecution,
+    TestResult,
+)
+from test_observer.data_access.models_enums import (
+    FamilyName,
+    TestExecutionReviewDecision,
+    TestExecutionStatus,
+    TestResultStatus,
+)
+from test_observer.data_access.repository import get_or_create
+from test_observer.data_access.setup import get_db
+
+from .logic import delete_previous_results
+from .models import C3TestResult, C3TestResultStatus, EndTestExecutionRequest
+
+router = APIRouter()
+
+
+@router.put("/end-test")
+def end_test_execution(request: EndTestExecutionRequest, db: Session = Depends(get_db)):
+    test_execution = _find_related_test_execution(request, db)
+
+    if test_execution is None:
+        raise HTTPException(status_code=404, detail="Related TestExecution not found")
+
+    delete_previous_results(db, test_execution)
+    _store_test_results(db, request.test_results, test_execution)
+
+    has_failures = test_execution.has_failures
+
+    test_execution.status = (
+        TestExecutionStatus.FAILED if has_failures else TestExecutionStatus.PASSED
+    )
+
+    prev_test_execution = _get_previous_test_execution(db, test_execution)
+
+    if (
+        prev_test_execution
+        and not has_failures
+        and prev_test_execution.is_approved
+        and _ran_all_previously_run_cases(prev_test_execution, test_execution)
+    ):
+        test_execution.review_decision = [
+            TestExecutionReviewDecision.APPROVED_ALL_TESTS_PASS
+        ]
+
+    if request.c3_link is not None:
+        test_execution.c3_link = request.c3_link
+
+    db.commit()
+
+
+def _ran_all_previously_run_cases(
+    prev_test_execution: TestExecution, test_execution: TestExecution
+) -> bool:
+    prev_test_cases = {
+        tr.test_case.name
+        for tr in prev_test_execution.test_results
+        if tr.status != TestResultStatus.SKIPPED
+    }
+
+    test_cases = {
+        tr.test_case.name
+        for tr in test_execution.test_results
+        if tr.status != TestResultStatus.SKIPPED
+    }
+
+    return prev_test_cases.issubset(test_cases)
+
+
+def _get_previous_test_execution(
```

Comment:
Could we potentially reuse some of the code from the `get_previous_test_executions_query` defined in `backend/test_observer/controllers/test_executions/logic.py`. I remember implementing the similar functionality for fetching the historic test results for a test execution.

Reply:
I thought about this as it is very similar functionality as you mentioned. But it doesn't seem possible to reuse that logic as is, and modifying it to serve both use cases seemed challenging for me (and I worried it would make the code less readable). But if you have an idea of how we can merge the two use cases without over complicating things I would definitely be interested.

Reply:
Wouldn't it be enough to call the `get_previous_test_executions_query` method with `.limit(1)` instead of `.limit(PREVIOUS_TEST_RESULT_COUNT)`. Although that will return just the ID of the test execution, so we would need one extra query to fetch the actual test execution.

This will hurt the performance slightly as we will be making one extra database query and maybe over complicating the logic.

I don't mind either approaches, we can also keep the logic for these two functionalities separate, as they do have some differences as well.

Reply:
I agree with you, even though we might be able to reuse a little bit of code, we will still need to write a few adapters and methods that will make it even more difficult to understand.

Let's keep it as it is and we can work on uniting the logic if we need to add more layers of complexity on top of it.